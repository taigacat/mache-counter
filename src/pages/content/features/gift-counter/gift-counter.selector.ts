import {createSelector} from 'reselect';
import {RootState} from '../../rootState.type';

export const giftCounterSelector = (state: RootState) => state.gifts;

export const giftSummerySelector = createSelector(giftCounterSelector, (gifts) => {
  const giftMap = gifts.reduce((acc, gift) => {
    if (acc[gift.name]) {
      acc[gift.name] += gift.count;
    } else {
      acc[gift.name] = gift.count;
    }
    return acc;
  }, {} as { [key: string]: number });

  return Object.entries(giftMap)
    .sort(([, countA], [, countB]) => countB - countA)
    .map(([name, count]) => ({name, count}));
});
