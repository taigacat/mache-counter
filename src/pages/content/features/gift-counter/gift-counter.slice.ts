import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { ChromeExtensionMessage } from '../../../../events/chrome-extension-message';
import { Gift, IndexedGift } from '../../../../models/Gift';
import { RootState } from '../../rootState.type';

export type GiftState = {
  gifts: { [key: string]: number };
  allGifts: IndexedGift[];
  sent?: boolean;
};

const initialState: GiftState = {
  gifts: {},
  allGifts: [],
  sent: false,
};

const giftCounterSlice = createSlice({
  name: 'gifts',
  initialState,
  reducers: {
    /**
     * Add gifts to the state
     * @param state current state
     * @param action action with gifts to add
     */
    add(state, action) {
      const payload = action.payload as Gift[];
      if (!payload || payload.length === 0) {
        return;
      }

      state.gifts = payload.reduce(
        (prev: { [name: string]: number }, gift: Gift) => {
          const current_gift = prev[gift.name];
          return {
            ...prev,
            [gift.name]: current_gift ? current_gift + gift.count : gift.count,
          };
        },
        state.gifts,
      );

      const count = state.allGifts.length;
      const indexedGifts = payload.map((gift, index) => ({
        ...gift,
        index: count + index,
      }));
      state.allGifts = [...state.allGifts, ...indexedGifts];
    },
    /**
     * Update gifts in the state
     * @param state current state
     * @param action action with gifts to update
     * @deprecated use add instead for performance reasons
     */
    update(state, action) {
      const payload = (action.payload || []) as Gift[];
      state.gifts = payload.reduce(
        (prev: { [name: string]: number }, gift: Gift) => {
          const current_gift = prev[gift.name];
          return {
            ...prev,
            [gift.name]: current_gift ? current_gift + gift.count : gift.count,
          };
        },
        {},
      );

      const indexedGifts = payload.map((gift, index) => ({ ...gift, index }));
      state.allGifts = [...indexedGifts];
    },
    /**
     * Set sent flag to true
     */
    sent(state, action) {
      state.sent = true;
    },
  },
});

export const sendGiftAsync = createAsyncThunk(
  'gifts/sendGiftAsync',
  async (diff: Gift[], thunkAPI) => {
    const state = thunkAPI.getState() as RootState;
    const metadata = state.app.metadata;
    const { allGifts, sent } = state.gift;
    ChromeExtensionMessage.sendEvent({
      metadata,
      event: 'gift',
      data: {
        ...(sent ? {} : { all: allGifts }),
        diff: diff.map((gift, index) => ({
          ...gift,
          index: allGifts.length - diff.length + index,
        })),
      },
    }).then((response) => {
      console.log(response);
      if (response.status === 'success') {
        thunkAPI.dispatch(giftAction.sent(undefined));
      }
    });
  },
);

export const giftAction = giftCounterSlice.actions;

export default giftCounterSlice;
