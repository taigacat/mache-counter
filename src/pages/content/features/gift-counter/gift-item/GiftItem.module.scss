.gift-item {
  display: flex;
  width: 100%;
  cursor: default;
  border-bottom: 1px solid lightgray;

  > div {
    padding: 0 5px;
    text-align: right;

    &:first-child {
      flex-grow: 1;
      padding-right: 15px;
    }

    &:last-child {
      width: 50px;
    }
  }

  &:hover {
    background-color: rgba(0, 0, 0, 0.1);
  }
}
