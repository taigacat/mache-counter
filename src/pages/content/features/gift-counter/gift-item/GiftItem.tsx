import React from 'react';
import {Gift} from '../../../../../models/Gift';
import styles from './GiftItem.module.scss';

/**
 * Represents a Prop object.
 * @interface
 */
interface Prop {
  gift: Gift;
}

/**
 * GiftItem component.
 *
 * @component
 *
 * @param {Object} props - The component props.
 * @param {Object} props.gift - Information about the gift.
 *
 * @returns {JSX.Element} The rendered GiftItem component.
 */
export const GiftItem: React.FC<Prop> = ({gift}) => {
  return (
    <div className={styles['gift-item']}>
      <div>{gift.name}</div>
      <div>{gift.count}</div>
    </div>
  );
}
