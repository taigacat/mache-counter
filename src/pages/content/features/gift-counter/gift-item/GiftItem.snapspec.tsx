import {GiftItem} from './GiftItem';
import {render} from '@testing-library/react';

describe('GiftItem snapshot', () => {
  it('should match snapshot', () => {
    // Arrange
    const gift = {
      name: 'Test Gift',
      count: 4,
    };

    // Act
    const {container} = render(
      <GiftItem gift={gift}/>
    );

    // Assert
    expect(container).toMatchSnapshot();
  });
});
