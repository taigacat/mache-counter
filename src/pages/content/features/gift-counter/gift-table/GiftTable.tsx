import React from 'react';
import Draggable from 'react-draggable';
import { useSelector } from 'react-redux';
import { giftSummerySelector } from '../gift-counter.selector';
import { GiftItem } from '../gift-item/GiftItem';
import styles from './GiftTable.module.scss';

export const GiftTable: React.FC = () => {
  const gifts = useSelector(giftSummerySelector);
  const windowHeight = window.innerHeight;
  return (
    <Draggable defaultPosition={{ x: 10, y: windowHeight - 210 }}>
      <div className={styles['gift-box']}>
        <div className={styles['gift-header']}></div>
        <div className={styles['gift-table']}>
          {gifts.map((gift) => (
            <GiftItem gift={gift} key={gift.name} />
          ))}
        </div>
      </div>
    </Draggable>
  );
};
