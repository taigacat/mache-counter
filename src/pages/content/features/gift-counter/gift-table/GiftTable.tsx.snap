// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GiftTable snapshot should match snapshot 1`] = `
<div>
  <div
    class="gift-table"
  >
    <div
      class="gift-item"
    >
      <div>
        Test Gift 1
      </div>
      <div>
        4
      </div>
    </div>
  </div>
</div>
`;
